crates/rustrace/src/main.rs

```diff
@@ -0,0 +1,69 @@
+use anyhow::{anyhow, bail, Result};
+use clap::Parser;
+use fork::{fork, Fork};
+use nix::{
+    errno::Errno,
+    sys::{
+        ptrace::{self, traceme, Options},
+        wait::waitpid,
+    },
+    unistd::Pid,
+};
+use syscalls::Sysno;
+
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    /// Command to trace
+    cmd: String,
+}
+
+fn main() -> Result<()> {
+    let args = Args::parse();
+
+    match fork() {
+        Ok(Fork::Parent(child)) => trace_child(child),
+        Ok(Fork::Child) => run_command_to_trace(&args.cmd),
+        Err(e) => Err(anyhow!("{e}")),
+    }
+}
+
+fn trace_child(child: i32) -> Result<()> {
+    let child = Pid::from_raw(child);
+    waitpid(Some(child), None)?;
+    // Jail the child
+    ptrace::setoptions(child, Options::PTRACE_O_EXITKILL)?;
+
+    loop {
+        // Stop on syscall invocation
+        ptrace::syscall(child, None)?;
+        waitpid(Some(child), None)?;
+
+        let registers = ptrace::getregs(child)?;
+        let syscall: Sysno = (registers.orig_rax as i32).into();
+        print!(
+            "{syscall}({}, {}, {}, {}, {}, {})",
+            registers.rdi, registers.rsi, registers.rdx, registers.r10, registers.r8, registers.r9
+        );
+
+        // Run the syscall
```

Comment:
It would be nice to explain here that `ptrace::syscall(); waitpid();` combination catches syscall entry as well as return and hence need to be called twice within the loop.

```suggestion
        // Let the syscall run and intercept on its return
```

Reply:
For sure, I was also confused by that so it won't hurt to add some context.